### PR TITLE
feat(serverless compaction): add retry strategy for grpc proxy client

### PR DESCRIPTION
**Pull request by wcy-fdu**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/rpc_client/src/compactor_client.rs [hunk 1]
```diff
@@ -82,56 +89,147 @@ impl GrpcCompactorProxyClientCore {
...
-        todo!()
+    async fn recreate_core(&self) {
+        tracing::info!("GrpcCompactorProxyClient rpc transfer failed, try to reconnect");
+        let channel = self.connect_to_endpoint().await;
+        let mut core = self.core.write().await;
+        *core = GrpcCompactorProxyClientCore::new(channel);
+    }
+
+    async fn connect_to_endpoint(&self) -> Channel {
+        let endpoint_str: &'static str = Box::leak(self.endpoint.clone().into_boxed_str());
```
Comment (Li0k) [inline src/rpc_client/src/compactor_client.rs:109]: How is the leak recycled here?

File: src/rpc_client/src/compactor_client.rs [hunk 2]
```diff
@@ -82,56 +89,147 @@ impl GrpcCompactorProxyClientCore {
...
+        let endpoint = Endpoint::from_static(endpoint_str);
```
Comment (arkbriar) [inline src/rpc_client/src/compactor_client.rs:110]: Shall we use `[from_shared](https://docs.rs/tonic/latest/tonic/transport/channel/struct.Endpoint.html#method.from_shared)` so we don't have to manually leak the endpoint?

Comment (wcy-fdu) [reply src/rpc_client/src/compactor_client.rs:110]: Make sense.

**Review by arkbriar — APPROVED** [review]
LGTM!

cc. @hzxa21 for another look.

**Review by Li0k — APPROVED** [review]
LGTM

**Pull request merged**